File: ctest/test_DGETRF.c
/* Test for LAPACKE_dgetrf - LU factorization */
#include <stdlib.h>
#include <stdio.h>
#include <math.h>
#include "lapacke.h"

/* Auxiliary routines prototypes (from lapacke_example_aux.c) */
extern void print_matrix_colmajor(char* desc, lapack_int m, lapack_int n, double* mat, lapack_int ldm);
extern void print_matrix_rowmajor(char* desc, lapack_int m, lapack_int n, double* mat, lapack_int ldm);
extern void print_vector(char* desc, lapack_int n, lapack_int* vec);

#define N 4
#define LDA N

int main() {
    lapack_int m = N, n = N, lda = LDA, info;
    lapack_int ipiv[N];

    /* Test matrix (column-major storage) */
    double a[LDA*N] = {
        6.80, -2.11,  5.66,  5.97,
       -6.05, -3.30,  5.36, -4.44,
       -0.45,  2.58, -2.70,  0.27,
        8.32,  2.71,  4.35, -7.17
    };

    printf("=== LAPACKE_dgetrf Test (Column-Major) ===\n");
    print_matrix_colmajor("Entry Matrix A", m, n, a, lda);

    /* Call LAPACKE_dgetrf */
    info = LAPACKE_dgetrf(LAPACK_COL_MAJOR, m, n, a, lda, ipiv);

    if (info > 0) {
        printf("U(%d,%d) is exactly zero - matrix is singular\n", info, info);
        return 1;
    } else if (info < 0) {
        printf("Parameter %d had an illegal value\n", -info);
        return 1;
    }

    printf("LAPACKE_dgetrf completed successfully\n");
    print_matrix_colmajor("LU factorization", m, n, a, lda);
    print_vector("Pivot indices", n, ipiv);

    /* Test row-major interface */
    printf("\n=== LAPACKE_dgetrf Test (Row-Major) ===\n");

    /* Row-major version of same matrix */
    double a_row[LDA*N] = {
        6.80, -6.05, -0.45,  8.32,
       -2.11, -3.30,  2.58,  2.71,
        5.66,  5.36, -2.70,  4.35,
        5.97, -4.44,  0.27, -7.17
    };
    lapack_int ipiv_row[N];

    print_matrix_rowmajor("Entry Matrix A (row-major)", m, n, a_row, lda);

    info = LAPACKE_dgetrf(LAPACK_ROW_MAJOR, m, n, a_row, lda, ipiv_row);

    if (info > 0) {
        printf("U(%d,%d) is exactly zero - matrix is singular\n", info, info);
        return 1;
    } else if (info < 0) {
        printf("Parameter %d had an illegal value\n", -info);
        return 1;
    }

    printf("LAPACKE_dgetrf (row-major) completed successfully\n");
    print_matrix_rowmajor("LU factorization (row-major)", m, n, a_row, lda);
    print_vector("Pivot indices (row-major)", n, ipiv_row);

    printf("\nAll DGETRF tests PASSED\n");
    return 0;
}

File: ftest/test_dgetrf.f90
! Test program for DGETRF (LU factorization)
program test_dgetrf
  implicit none

  integer, parameter :: m = 3, n = 3
  double precision :: a(m,n), a_orig(m,n)
  integer :: ipiv(min(m,n)), info

  ! Matrix A = [[1, 2, 3], [4, 5, 6], [7, 8, 10]]
  a(1,1) = 1.0d0; a(1,2) = 2.0d0; a(1,3) = 3.0d0
  a(2,1) = 4.0d0; a(2,2) = 5.0d0; a(2,3) = 6.0d0
  a(3,1) = 7.0d0; a(3,2) = 8.0d0; a(3,3) = 10.0d0

  a_orig = a

  print *, "Testing DGETRF..."
  print *, "Matrix A:"
  print *, a(1,:)
  print *, a(2,:)
  print *, a(3,:)

  ! Call DGETRF
  call dgetrf(m, n, a, m, ipiv, info)

  if (info == 0) then
    print *, "DGETRF succeeded!"
    print *, "LU factorization:"
    print *, a(1,:)
    print *, a(2,:)
    print *, a(3,:)
    print *, "Pivot:", ipiv
  else if (info > 0) then
    print *, "Matrix is singular, U(", info, ",", info, ") = 0"
    stop 1
  else
    print *, "DGETRF failed with info =", info
    stop 1
  end if

  print *, "Test PASSED"
end program test_dgetrf
